### ISSUE TITLE
Please consider transpiling to color(display-p3) when available

### ISSUE THREAD

**Issue by LeaVerou** [csstools/postcss-lab-function#7]
Right now, `lab()` and `lch()` are converted to `rgb()` colors, which per-spec is restricted to the sRGB gamut. Safari supports the `color(display-p3 r g b)` notation from CSS Color 4, which can access the entire P3 gamut. Please consider transpiling to both and using the cascade (or `@supports`) to utilize the one the browser supports. You can find conversion math in https://css.land/lch

**Comment by jonathantneal** [comment]
Thanks for the recommendation, for pointing me to that project, and for posting the math on GitHub, @LeaVerou.

I think what I need to do is:

- [ ] Copy the relevant math from [`lch.js`](https://github.com/LeaVerou/css.land/blob/master/lch/lch.js) into [convert-colors](https://github.com/jonathantneal/convert-colors) ([docs](https://jonneal.dev/convert-colors/)) ([example math with citation](https://github.com/jonathantneal/convert-colors/blob/master/src/hwb-hsv.js#L24-L44)).
- [ ] Update this plugin to support those p3 transforms.
- [ ] Update or change this plugin’s [`preserve`](https://github.com/csstools/postcss-lab-function#preserve) option to accommodate the differing possibilities (as browsers gain support).

**Comment by LeaVerou** [comment]
Do note that a lot of the math is written by @svgeesus and resides in the csswg-drafts repo. css.land/lch links to it. We are working towards writing a more OOP library without the math.js dependency and releasing it separately, but no ETA for that. I will try to update this issue as things progress on that front, I imagine even just getting rid of the dependency will help and that can be done much sooner.

**Comment by LeaVerou** [comment]
Heads up, I just pushed a change that gets rid of the math.js dependency and replaces it with a much smaller (~40 lines) `multiplyMatrices()` function, so the conversion logic is now self-contained!

**Comment by jonathantneal** [comment]
@LeaVerou, thanks for the update. Here’s the update on my end.

My work was temporarily blocked by a bug introduced in v3 of [PostCSS Values Parser](https://github.com/shellscape/postcss-values-parser/pull/109), which is now fixed in [v3.2.0](https://github.com/shellscape/postcss-values-parser/releases/tag/v3.2.0).

After investigating the `LCH_to_P3()` function in [utilities.js#L77-L89](https://github.com/w3c/csswg-drafts/blob/master/css-color-4/utilities.js#L77-L89) from [`csswg-drafts/css-color-4`](https://github.com/w3c/csswg-drafts/tree/master/css-color-4), I see that the conversion goes like this:

- Transform `LCH` into `LAB` in [conversions.js#L342-L349](https://github.com/w3c/csswg-drafts/blob/master/css-color-4/conversions.js#L342-L349)
- Transform `LAB` into `XYZ` in [conversions.js#L308-L330](https://github.com/w3c/csswg-drafts/blob/master/css-color-4/conversions.js#L308-L330).
- Transform D50 `XYZ` to D65 `XYZ` in [conversions.js#L275-L284](https://github.com/w3c/csswg-drafts/blob/master/css-color-4/conversions.js#L275-L284).
- Transform D65 `XYZ` to linear-light P3 `RGB` in [conversions.js#L90-L99](https://github.com/w3c/csswg-drafts/blob/master/css-color-4/conversions.js#L90-L99).
- Transform linear-light P3 `RGB` to gamma-corrected P3 `RGB` in [conversions.js#L20-L31](https://github.com/w3c/csswg-drafts/blob/master/css-color-4/conversions.js#L20-L31).

And it looks like I am already doing this.

- Transform `LCH` into `LAB` in [lab-lch.js#L38-L46](https://github.com/jonathantneal/convert-colors/blob/master/src/lab-lch.js#L38-L46).
- Transform `LAB` into `XYZ` in [lab-xyz.js#L17-L42](https://github.com/jonathantneal/convert-colors/blob/master/src/lab-xyz.js#L17-L42).
- Transform D50 `XYZ` to D65 `XYZ` in [lab-xyz.js#L33-L38](https://github.com/jonathantneal/convert-colors/blob/master/src/lab-xyz.js#L33-L38).
- Transform D65 `XYZ` to linear-light P3 `RGB` in [rgb-xyz.js#L44-L48](https://github.com/jonathantneal/convert-colors/blob/master/src/rgb-xyz.js#L44-L48).
- Transform linear-light P3 `RGB` to gamma-corrected P3 `RGB` in [rgb-xyz.js#L50-L52](https://github.com/jonathantneal/convert-colors/blob/master/src/rgb-xyz.js#L50-L52).

I believe this means I am producing P3 RGB colors. Does this mean I can keep my logic, but that I also need to change the color conversion library to produce these lower gamut sRGB colors?

**Comment by LeaVerou** [comment]
That's such a great outline of the conversion algorithm!

(Keep in mind that I haven't used the plugin, and this is based on a quick skim of your code and tests, so it's entirely possible I missed something)

You may be calculating P3 coordinates, but I believe the output eventually is an `rgb()` function, right? `rgb()` **cannot access colors outside the sRGB gamut**. Even if you provide values below 0 or above 100%/255 to accomplish this, they will merely be clipped.
To actually access P3 colors today, the only option is to use the `color(display-p3)` syntax. This would need to be provided in addition to any `rgb()` values, since its support is limited. To reduce amount of code, you could only do that when the color is outside the sRGB gamut, no need to do this if it's not.

In addition, for generating the `rgb()` values when the color is outside sRGB, it would be a good idea to shrink it to the sRGB gamut yourself instead of outputting an `rgb()` color with values out of range, since then the browser just clips them, whereas it's a far better algorithm to reduce chroma until the color is within gamut, as done [here](https://github.com/LeaVerou/css.land/blob/master/lch/lch.js#L35) (thanks to a lovely PR by @tabatkins)

**Comment by svgeesus** [comment]
> After investigating the LCH_to_P3() function in utilities.js#L77-L89 from csswg-drafts/css-color-4, I see that the conversion goes like this:

Yes. These are also the steps listed [in the specification](https://drafts.csswg.org/css-color-4/#lab-to-predefined).

As @LeaVerou says, the P3 values should be used inside `color(display-p3 r g b)` because `rgb(r g b)` means the sRGB colorspace.

**Comment by matheusmurden** [comment]
Any updates on this?

**Comment by romainmenke** [comment]
@svgeesus Have been looking into this request and there seems to be some discrepancy in colors when using the functions listed above.

Left are processed from `lab` to `color` and `rgb` as fallback.
Right is the original `lab` value.

These are direct results from running `gam_P3(XYZ_to_lin_P3(D50_to_D65(Lab_to_XYZ(labValues))));`
Where `labValues` is an array containing (in this case) `[45, -128, -128]`.

<img width="142" alt="Screenshot 2021-11-14 at 23 34 41" src="https://user-images.githubusercontent.com/11521496/141701283-730ac9b8-909a-43fe-97ca-88c3daaa5dae.png">


Color as `lab()` :

`lab(45% -128 -128)`

<img width="688" alt="Screenshot 2021-11-14 at 23 29 55" src="https://user-images.githubusercontent.com/11521496/141701188-ebcc0ca4-cbf4-4c83-9875-4eafb26a1c7a.png">

Color as `color(display-p3 ...)` :

`color(display-p3 -0.8084798483397213 0.5531899229209936 1.2454753974760675)`

<img width="707" alt="Screenshot 2021-11-14 at 23 30 01" src="https://user-images.githubusercontent.com/11521496/141701191-b4ae2587-9dec-4a55-becb-a05f5013b2b7.png">

Color as `rgb()` :

`rgb(0, 150, 255)`

<img width="729" alt="Screenshot 2021-11-14 at 23 30 12" src="https://user-images.githubusercontent.com/11521496/141701194-abdb31f3-4bde-4cb2-96c4-c91cd8764df2.png">

------

Some color ranges seem affected, others seem fine :

<img width="125" alt="Screenshot 2021-11-14 at 23 37 57" src="https://user-images.githubusercontent.com/11521496/141701428-71153a88-edc9-4b45-92f3-a0e7a69ae875.png">

------

Am I holding it wrong? :)

----

Update : 

also opened an issue here to see were it goes wrong :
https://github.com/w3c/csswg-drafts/issues/6816

**Comment by svgeesus** [comment]
As mentioned in [the other issue you opened](https://github.com/w3c/csswg-drafts/issues/6816), you are specifying extreme out of gamut colors, which are not only outside the range of `rec2020` but actually not physically realizable. You then see differences in gamut mapping strategies.

For example [`lab(45% -128 -128`)](https://colorjs.io/apps/convert/?color=lab(45%25%20-128%20-128)&precision=4) is `color(srgb -0.91 0.587 1.295)` or `color(rec2020 -1.51 0.483 1.27)` notice the component values well outside the 0..1 range.

**Comment by romainmenke** [comment]
Thank you for confirming here.

I think the "out of gamut" colors would need some handling from the perspective of the plugin.
Have to give it more thought on what output would be least problematic.

Not doing anything is what happens in the examples I referenced in the other issue.
Min/Max between 0 and 255 is used here but also leads to issues.

**Comment by svgeesus** [comment]
Oh, certainly out of gamut colors need to be handled. My point is that such colors make poor unit tests, because you are testing multiple things at once.

**Comment by svgeesus** [comment]
There is an existing CSS Color 4 issue [Define gamut mapping](https://github.com/w3c/csswg-drafts/issues/5191). We recently resolved to do this using Chroma reduction in OKLCH.

**Comment by romainmenke** [comment]
Aha! That sounds really interesting.
Having the correct implementation in this plugin would be great.

> My point is that such colors make poor unit tests, because you are testing multiple things at once.

True.
Would make much more sense to filter out any color that is out of gamut.
I will try to figure this out soon so that we at least know that in gamut conversions between `lab`/`lch` -> `srgb`/`display-pr` are correct.

That was also the reason for me to comment here and open that other issue. I am mainly looking for a good source of truth so that I know which bits of the code in this plugin are working correctly. Just iterating everything and displaying it was my naive attempt to do this.

**Comment by romainmenke** [comment]
Gammut mapping was merged into spec : https://github.com/w3c/csswg-drafts/commit/d51a437cc8be69b864a11f0cf9b8e6f5f6532973

Note to self : check if this contains enough info to figure out how to work this into the plugin

**Comment by svgeesus** [comment]
@romainmenke I would be very interested to hear detailed comments on anything you find vague, confusing or under-specified.

I'm working on a pseudo-code algorithm for the spec right now.

**Comment by romainmenke** [comment]
@svgeesus Just read through it and it is all very clear to me.
I am looking forward to the pseudo-code algorithm!

Just to check if I read everything correctly (if I didn't, it says more about me than about your explanation)

1. start with a color in a large color space
2. covert to srgb naively
2.1. if in gamut, return the color (no values smaller than 0 or larger than 1)
2.2. if out of gamut continue
3. use something like binary search to find the smallest change (via `deltaEOK`) for the following
3.1. convert to OKLCH
3.2. reduce chroma value by X
3.3. convert to srgb
3.4. check if in gamut and continue search
4. return srgb result of smallest change

----

edit : https://drafts.csswg.org/css-color-4/#binsearch

**Comment by romainmenke** [comment]
@svgeesus There is one thing that might be a bit confusing :

https://drafts.csswg.org/css-color-4/#binsearch

> 6. let clip(color)| be a function which converts color to destination, converts all negative components to zero, converts all components greater that one to one, and returns the result

Any color returned here would be in the destination color space.

> 9.4. otherwise, if inGamut(current) is false carry out these steps:
> 9.4.1. set clipped to clip(current)
> 9.4.2. set E to delta(clipped, current)
> 9.4.3. if E < JND convert clipped to destination and return it as the gamut mapped color

If we convert the clipped values to destination it would be converted twice.

Is it also correct that `clipped` and `current` are passed to `deltaEOK` but both are in different color spaces?

**Comment by romainmenke** [comment]
Maybe needs to be :

> 9.4. otherwise, if inGamut(current) is false carry out these steps:
> 9.4.1. set clipped to clip(current)
> 9.4.2. let clipped_OKLCH be clipped converted to the OKLCH color space
> 9.4.3. set E to delta(clipped_OKLCH, current)
> 9.4.4. if E < JND return clipped as the gamut mapped color

**Comment by romainmenke** [comment]
@svgeesus I played around with this and the results are interesting but there are still some bugs in my code.

Both my conversions and the colors shown in Safari with `background-color: lab(...)` look wrong in some out of gamut cases. _My expectations might be incorrect for how colors are rendered when so far outside of a color space._

When out of gamut Safari almost never matches the conversion.
But that they display `lab(0% -127 -127)` as a bright blue makes me think they don't do any gamut mapping?

_I am comparing with Safari because it shipped there but it's really comparing apples with oranges without knowing if the oranges are in fact oranges._

------

Gamut mapping code :

https://github.com/romainmenke/lab-lch-display-p3/blob/main/css-color-4/map-gamut.ts

I found that `chroma` can become negative and added a shortcut for it.
Also found that the amount used to reduce `chroma` can become very small leading to an infinite loop

LAB to sRGB conversion :

_LAB to p3 in the same file is used in the visualisation to show which are outside of the p3 color space_

https://github.com/romainmenke/lab-lch-display-p3/blob/main/css-color-4/convert-lab-to-srgb.ts#L26

------

Also tried to make a better visualisation of the colors :

https://romainmenke.github.io/lab-lch-display-p3/lab-interactive.html

<img width="880" alt="Screenshot 2021-12-12 at 13 37 44" src="https://user-images.githubusercontent.com/11521496/145712568-d6d49eaa-0add-4815-bc01-3a23551238e9.png">

**Comment by svgeesus** [comment]
> Just to check if I read everything correctly (if I didn't, it says more about me than about your explanation)

No actually, this is the first time I am writing this sort of html5-like english-language (or BASIC with GOTO, it looks like to me) algorithm so I am a bit clumsy with trying to represent for and while loops in this way.

I'm also used to working in color.js where all the intermediate steps get done for you, while this algorithm needs to state them explicitly.

> return srgb result of smallest change

Kind of. Return an srgb result which will be either:

- a value which fell just inside the gamut, or
- the clipped version of a value which fell just outside the gamut

**Comment by svgeesus** [comment]
> Any color returned here would be in the destination color space.

Yes.

> Is it also correct that clipped and current are passed to deltaEOK but both are in different color spaces?

Oh right, I need to say so explicitly in the definition of `delta`, or add an extra conversion step:

> 9.4.1. set clipped to clip(current)
9.4.2. let clipped_OKLCH be clipped converted to the OKLCH color space
9.4.3. set E to delta(clipped_OKLCH, current)
9.4.4. if E < JND return clipped as the gamut mapped color

Nice. Yes, exactly.

**Comment by svgeesus** [comment]
> When out of gamut Safari almost never matches the conversion.
But that they display lab(0% -127 -127) as a bright blue makes me think they don't do any gamut mapping?

If the shipping version of Safari matched an algorithm I put in the spec 6 days ago then I would want to know where their time machine was.
Yeah, since the spec was silent on gamut mapping until recently, I think they just do a simple clip.

**Comment by romainmenke** [comment]
> If the shipping version of Safari matched an algorithm I put in the spec 6 days ago then I would want to know where their time machine was.

😂

> the first time I am writing this sort of html5-like english-language (or BASIC with GOTO, it looks like to me) 

I am used to reading tc39 specs but writing an algorithm in this style is not easy.

-----------

Generally I think the current result I am seeing are pretty ok.
Is there a set of values (both in and out of gamut) that I could use to verify the current code?

Started a PR for this with the most relevant bit here : https://github.com/csstools/postcss-plugins/blob/f3ccb446e5fcc65209a868ede721161a653e28ab/plugins/postcss-lab-function/src/css-color-4/map-gamut.ts#L9-L29

Also updated the visualisation : https://romainmenke.github.io/lab-lch-display-p3/lab-interactive.html

There is this weird blotch around `lab(x, -127, -127)`
But it is way out of gamut and I don't know if this is expected and ok, or unexpected and that there is a bug somewhere?

<img width="943" alt="Screenshot 2021-12-13 at 18 21 52" src="https://user-images.githubusercontent.com/11521496/145859235-c4504f06-6770-4f91-987d-b896fe3c0156.png">

**Comment by svgeesus** [comment]
> I found that chroma can become negative and added a shortcut for it.
Also found that the amount used to reduce chroma can become very small leading to an infinite loop

I can easily believe that the chroma reductions can become very small and the effect invisible. A better terminating condition would be welcome.

I don't see how there can be a negative chroma, except perhaps from round-trip round-off errors with neutral colors and OKLCH -> sRGB -> OKLCH?

The weird blotch is troubling, especially the  sharp discontinuity, but that color is likely way outside the spectral locus (ie not just out of gamut of a particular display, but not an actual color at all)

Here is lab(50% -127 -127) on a uv chromaticity diagram. It is the big teal circle. The larger gamut is rec2020, the smaller is sRGB and the tongue shape is the spectral locus.

![image](https://user-images.githubusercontent.com/2506926/145867430-310db8d0-436a-47e3-8419-e97b23b07a66.png)

**Comment by romainmenke** [comment]
>I can easily believe that the chroma reductions can become very small and the effect invisible. A better terminating condition would be welcome.

I intend to plot these to see if there is any pattern, might have time for this in the next few days.

I found that there are often 2-3 iterations which no longer have any effect on the actual sRGB color. (considering that decimals in `rgb(...)` are rounded by browsers)

>I don't see how there can be a negative chroma, except perhaps from round-trip round-off errors with neutral colors and OKLCH -> sRGB -> OKLCH?

That was a bug on my end and this has been fixed.

>The weird blotch is troubling, especially the sharp discontinuity, but that color is likely way outside the spectral locus (ie not just out of gamut of a particular display, but not an actual color at all)

Yes it's definitely outside the spectral locus. That I display this as a square also explains the inverted curve at the edge of the blotch.

Can I ignore it? Or should we try to find out why there is this sharp edge and jump in luminosity.

**Comment by romainmenke** [comment]
@LeaVerou @svgeesus This repo has moved to [@csstools/postcss-plugins](https://github.com/csstools/postcss-plugins)

We would like to archive this repo which would make further discussion here impossible.
Is it ok if I ping you for feedback there?

**Comment by svgeesus** [comment]
> Is it ok if you ping you for feedback there?
Yes, of course
